includes/validation/class-amp-css-length.php

```diff
@@ -0,0 +1,142 @@
+<?php
+/**
+ * Class AMP_CSS_Length
+ *
+ * @package AMP
+ */
+
+/**
+ * Class AMP_CSS_Length
+ *
+ * Adapted from the <code>amp.validator.CssLength</code> class found in validator.js of the
+ * <a href="https://github.com/ampproject/amphtml">ampproject/amphtml</a> project.
+ *
+ * @version 1911070201440
+ * @link https://github.com/ampproject/amphtml/blob/1911070201440/validator/engine/validator.js#L3351
+ *
+ * @since 1.5
+ */
+class AMP_CSS_Length {
+
+	/**
+	 * Whether the value or unit is invalid. Note that passing
+	 * an empty value as `$attr_value` is considered valid.
+	 *
+	 * @var bool
+	 */
+	protected $is_valid = false;
+
+	/**
+	 * Whether the attribute value is set.
+	 *
+	 * @var bool
+	 */
+	protected $is_set = false;
+
+	/**
+	 * Whether the attribute value is 'auto'. This is a special value that
+	 * indicates that the value gets derived from the context. In practice
+	 * that's only ever the case for a width.
+	 *
+	 * @var bool
+	 */
+	protected $is_auto = false;
+
+	/**
+	 * Whether the attribute value is 'fluid'.
+	 *
+	 * @var bool
+	 */
+	protected $is_fluid = false;
+
+	/**
+	 * The numeric value.
+	 *
+	 * @var float
+	 */
+	protected $numeral = 0;
+
+	/**
+	 * The unit, 'px' being the default in case it's absent.
+	 *
+	 * @var string
+	 */
+	protected $unit = 'px';
+
+	/**
+	 * AMP_CSS_Length constructor.
+	 *
+	 * @param string $attr_value  Attribute value to be parsed.
+	 * @param bool   $allow_auto  Whether or not to allow the 'auto' value as a value.
+	 * @param bool   $allow_fluid Whether or not to allow the 'fluid' value as a value.
+	 */
+	public function __construct( $attr_value, $allow_auto, $allow_fluid ) {
+		if ( ! isset( $attr_value ) || '' === $attr_value ) {
+			$this->is_valid = true;
+			return;
+		}
+
+		$this->is_set = true;
+
+		if ( 'auto' === $attr_value ) {
+			$this->is_auto  = true;
+			$this->is_valid = $allow_auto;
+			return;
+		} elseif ( 'fluid' === $attr_value ) {
+			$this->is_fluid = true;
+			$this->is_valid = $allow_fluid;
+		}
+
+		$pattern = '/^(?<numeral>\d+(?:\.\d+)?)(?<unit>px|em|rem|vh|vw|vmin|vmax)?$/';
```

Comment:
I think letting it match any string of characters would defeat the purpose of the validation being done here. Also, the `%` wasn't included since its invalid according to the AMP validator:

https://github.com/ampproject/amphtml/blob/1911070201440/validator/engine/validator.js#L3407

Reply:
OK, you're right. Limiting to the recognized units is probably right. Nevertheless, we do need to account for `%` units in conversions for `fixed-height` and `fill` layouts. I would think that `AMP_CSS_Length` would be involved for parsing that out.

Reply:
Doesn't need to be. `AMP_Tag_And_Attribute_Sanitizer::adapt_dimensions_to_layout()` takes care of `%` units now. Anything else out of that norm should be removed.